### PR TITLE
Tooltips phase1

### PR DESCRIPTION
**Pull request by michneaionuc**
## Summary

This PR introduces the first end-to-end implementation of `ToolTip` as popup-based overlays, intended for early review before polishing behavior and API details.

Implemented so far:

- Added a `ToolTip` builtin element and runtime item support.
- Added compiler lowering for `ToolTip` into popup overlays.
- Introduced popup behavior classification (`PopupKind` / compiler-side `PopupWindowKind`) and wired it through compiler -> LLR -> codegen -> interpreter -> runtime.
- Marked tooltip popups explicitly at creation time (removed id-prefix heuristic).
- Runtime behavior for tooltip popups:
  - input-transparent (does not take pointer dispatch)
  - no focus stealing
  - unclamped placement request in popup placement logic
- Added initial interpreter test coverage for tooltip popup overlay behavior.

## Current behavior / limitations

- Tooltip position is currently derived from mouse coordinates (`mouse-x` / `mouse-y`).
- On backends without native popup window support (child-window fallback), tooltip rendering is still bounded by the parent window surface (cannot visually extend outside the main window).
- Positioning/collision handling is still basic (no polished anchor/preferred side/edge avoidance strategy yet).
- Styling and theming integration is still minimal and may need refinement.

## Follow-up work

Planned follow-ups after this initial review:

- Improve tooltip positioning strategy:
  - anchor-based placement (not cursor-only)
  - viewport/screen-aware collision handling
  - offsets and preferred side rules
- Clarify and stabilize API surface:
  - declarative positioning knobs
  - open/close behavior and timing controls
- Backend behavior improvements:
  - better native popup usage where available
  - define expected behavior for backends that use child-window fallback
- Expand tests:
  - positioning and edge/corner cases
  - backend-specific behavior expectations
  - interaction behavior (hover transitions, nested popups, etc.)

```
import { Button, VerticalBox } from "std-widgets.slint";

export component App {
    width: 360px;
    height: 280px;

    in-out property <string> tooltip-msg: "Save your changes";

    VerticalBox {
        padding: 24px;
        spacing: 16px;
        alignment: start;

        Text {
            text: "Hover the button — tooltip follows the pointer.";
            wrap: word-wrap;
        }

        TouchArea {
            width: 160px;
            height: 40px;

            Button {
                width: 100%;
                height: 100%;
                text: "Save";
                clicked => { debug("clicked"); 
                    popup.show();
                }
            }

            ToolTip {
                text: root.tooltip-msg;
            }
        }

        popup := PopupWindow {
            Text {
                text: root.tooltip-msg;
                color: red;
                wrap: word-wrap;
            }
        }
    }
}
```

### CODE REVIEW

File: internal/compiler/object_tree.rs [hunk 1]
```diff
@@ -824,6 +832,12 @@ pub struct Element {
     /// because the popup references it.
     pub has_popup_child: bool,
 
+    /// Semantic kind for elements that represent popup windows.
+    ///
+    /// Defaults to `Regular`; passes that synthesize specific popup kinds (such as tooltips)
+    /// should set this explicitly so later lowering does not rely on naming conventions.
+    pub popup_window_kind: PopupWindowKind,
```
Comment (tronical) [inline internal/compiler/object_tree.rs]: I think that if this is popup specific, it should be an `Option`. `PopupWindowKind::Regular` doesn't really make sense for a declared `Image { ... }` for example. If it's `None` then it's maybe clearer that this is not a popup.

File: internal/compiler/passes/lower_tooltips.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+//! Lowers `ToolTip { text: ... }` to an input-transparent popup overlay.
```
Comment (tronical) [inline internal/compiler/passes/lower_tooltips.rs:4]: It would be great if you could expand here a little what this gets lowered to, especially in terms of the logic when/how to show.

What I think is also absent is the logic to delay. Mouse triggered tooltips typically don't show immediately when hovering over their area but only after a delay of inactivity, right?

File: internal/compiler/passes/lower_tooltips.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            .values
+            .iter()
+            .position(|v| v == "no-auto-close")
+            .unwrap(),
+        enumeration: popup_close_policy_enum,
+    };
+
+    let mut tooltip_popup_id_counter: u32 = 0;
+    recurse_elem_including_sub_components_no_borrow(component, &(), &mut |elem, _| {
+        if elem.borrow().id.starts_with(TOOLTIP_POPUP_ID_PREFIX) {
```
Comment (tronical) [inline internal/compiler/passes/lower_tooltips.rs]: This looks like a recursion guard. Why is this needed? (what kind of markup ends up producing a recursion?)

File: internal/compiler/passes/lower_tooltips.rs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        }
+
+        let supports_hover = {
+            let elem_borrow = elem.borrow();
+            !matches!(elem_borrow.lookup_property(HAS_HOVER).property_type, Type::Invalid)
+                && !matches!(elem_borrow.lookup_property(MOUSE_X).property_type, Type::Invalid)
+                && !matches!(elem_borrow.lookup_property(MOUSE_Y).property_type, Type::Invalid)
+        };
+        if !supports_hover {
+            return;
```
Comment (tronical) [inline internal/compiler/passes/lower_tooltips.rs:426]: This feels a bit fragile. So if I use `ToolTip` as a child of something that doesn't support hover, then we're just silently going to do nothing? :)

If at the moment this only works as a child of TouchArea, then perhaps `TouchArea` needs to be taught that (see `Path` in `builtins.slint` that ensures that `ArcTo` for example can't be used outside). Then we can catch that situation earlier and produce an error.

If that shouldn't be part of this PR, then perhaps it should be part of the plan.

File: internal/core/window.rs [hunk 5]
```diff
@@ -1440,13 +1453,18 @@ impl WindowInner {
             .and_then(|x| x.create_popup(LogicalRect::new(position, size)))
         {
             None => {
-                let clip = LogicalRect::new(
-                    LogicalPoint::new(0.0 as crate::Coord, 0.0 as crate::Coord),
-                    self.window_adapter().size().to_logical(self.scale_factor()).to_euclid(),
-                );
+                // Should Tooltips extend past the window? If not, remove this block.
+                let clip_region = if popup_kind == PopupKind::Tooltip {
+                    None
```
Comment (tronical) [inline internal/core/window.rs]: Can you elaborate what this removal does and how it relates to the in-window (overlay) tooltips we're looking at?

File: internal/interpreter/eval.rs [hunk 6]
```diff
@@ -1872,6 +1872,19 @@ pub fn load_property(component: InstanceRef, element: &ElementRc, name: &str) ->
     load_property_helper(&ComponentInstance::InstanceRef(component), element, name)
 }
 
+pub(crate) fn lookup_rtti_str_key<'a, T>(
```
Comment (tronical) [inline internal/interpreter/eval.rs]: Why is this needed?

Normalization of properties should happen only right below the public API level (see call to `normalize_identifier` in `ComponentInstance`'s `get_property`).

File: tests/cases/elements/tooltip_popup_overlay.slint [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        y: 0px;
+        width: parent.width;
+        height: parent.height;
+
+        ToolTip {
+            text: "hello tooltip";
+        }
+    }
+}
+
```
Comment (tronical) [inline tests/cases/elements/tooltip_popup_overlay.slint:24]: This test should be extended to simulate mouse activity and verify that the tooltip was instantiated. Perhaps that verification could be done via an `init => {}` callback triggering activity - similar to how the popup window tests work?

**Review by tronical — COMMENTED** [review]
I think this is going into the right direction :-). But I've got a bunch of questions.

File: internal/compiler/object_tree.rs (hunk 1 above)
Comment (michneaionuc) [reply internal/compiler/object_tree.rs]: Good point, updated. Element.popup_window_kind is now Option<PopupWindowKind>, so non-popup elements are None, and popup elements are explicitly marked (Some(Regular) / Some(Tooltip)).

File: internal/compiler/passes/lower_tooltips.rs (hunk 2 above)
Comment (michneaionuc) [reply internal/compiler/passes/lower_tooltips.rs:4]: Updated this pass docs to describe what gets synthesized and the show/close flow.
I also implemented delayed show: hover starts/restarts a synthesized timer, timer trigger shows tooltip, hover leave stops timer and closes popup.

File: internal/compiler/passes/lower_tooltips.rs (hunk 3 above)
Comment (michneaionuc) [reply internal/compiler/passes/lower_tooltips.rs]: You’re right that this is a recursion/re-entry guard, but it’s pass-internal rather than source-markup recursion.
During traversal, this pass inserts synthesized tooltip popup nodes into the tree; those can be revisited in the same traversal.
I replaced the id-prefix-based skip with a structural guard: skip nodes where base_type == PopupWindow and popup_window_kind == Some(Tooltip).

Let me know if my understanding is correct here or if we can follow a better approach.

File: internal/compiler/passes/lower_tooltips.rs (hunk 4 above)
Comment (michneaionuc) [reply internal/compiler/passes/lower_tooltips.rs:426]: Agreed, silent no-op was fragile.
It now emits a compiler error when ToolTip is used under a parent that does not provide hover/mouse-position properties (currently TouchArea-like parents).  

Since this PR targets the feature branch, I plan to add support for tooltips on non-TouchArea visual elements in iteration 2 (likely via generated hover sensing, or other ways if you've got any ideas here), so this restriction is temporary for this phase.

File: internal/core/window.rs (hunk 5 above)
Comment (michneaionuc) [reply internal/core/window.rs]: For this phase I’m keeping tooltip placement aligned with current popup-style bounded placement.
I plan to implement dedicated tooltip placement in phase 2 (above/below/left/right with edge-aware selection), since that likely needs a fuller placement policy instead of the current popup-like behavior.

File: internal/interpreter/eval.rs (hunk 6 above)
Comment (michneaionuc) [reply internal/interpreter/eval.rs]: Removed.
I dropped interpreter-side RTTI key fallback normalization and switched back to direct RTTI key lookups, so normalization is no longer happening in this lower layer.

File: tests/cases/elements/tooltip_popup_overlay.slint (hunk 7 above)
Comment (michneaionuc) [reply tests/cases/elements/tooltip_popup_overlay.slint:24]: Extended.
The test now dispatches pointer movement, waits mocked elapsed time for tooltip delay, and verifies tooltip instantiation via ToolTip.init => { root.tooltip-created = true; }.

I also added tooltip_repeater_nested.slint to stress nested/repeater tooltip lowering and verify delayed instantiation there as well.

File: internal/compiler/builtins.slint [hunk 8]
```diff
@@ -153,6 +153,16 @@ export component TouchArea {
     //-default_size_binding:expands_to_parent_geometry
 }
 
+export component TooltipArea inherits Empty {
+    out property <bool> has-hover;
+    //-default_size_binding:expands_to_parent_geometry
```
Comment (ogoffart) [inline internal/compiler/builtins.slint:160]:
```suggestion
    //-default_size_binding:expands_to_parent_geometry
    //-is_internal
```

(So that user can't use it directly)

File: internal/compiler/builtins.slint [hunk 9]
```diff
@@ -153,6 +153,16 @@ export component TouchArea {
...
+}
+
+export component ToolTip inherits Empty {
+    in property <string> text;
+    in property <ToolTipPlacement> placement: top;
```
Comment (ogoffart) [inline internal/compiler/builtins.slint]: The tooltip usually appear wherever the mouse it, for me. I suggest we leave this out of the API now.

**Comment by michneaionuc** [comment]
Merged this as it's been the full implementation afetr all. It's easier to update the PR from the master branch directly. Gonna greate a PR to master soon.
